internal/middleware/websocketcors.go

```diff
@@ -0,0 +1,47 @@
+// Copyright 2024 Canonical.
+package middleware
+
+import (
+	"net/http"
+
+	"github.com/rs/cors"
+)
+
+// WebsocketCors provides middleware for handling CORS on websockets.
+type WebsocketCors struct {
+	cors *cors.Cors
+}
+
+// NewWebsocketCors returns a new WebsocketCors object.
+// If no allowedOrigins are provided, all origins will be allowed.
+func NewWebsocketCors(allowedOrigins []string) *WebsocketCors {
+	corsOpts := cors.New(cors.Options{
+		AllowedOrigins: allowedOrigins,
+	})
+	return &WebsocketCors{cors: corsOpts}
+}
+
+// Handler implements CORS validation for websocket handlers.
+// Any methods beside GET are rejected as a bad request.
+// If the origin is not in the allow list, a status forbidden is returned.
+func (c *WebsocketCors) Handler(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != http.MethodGet {
+			w.WriteHeader(http.StatusBadRequest)
+			_, _ = w.Write([]byte("unexpected method"))
+			return
+		}
+
+		// We can allow cross-origin requests for those requests that don't
```

Comment:
requests coming from juju clients what origin do they have? From your code I imagine nothing.

Reply:
Nothing, our tests prove as much, otherwise CLI requests would start failing. And as far as I understand the `origin` header is added by browsers.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin
https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name